SI-8790 detuplify match under -Xexperimental

adriaanm opened the pull request:
Optimize

```
(e1,..., eN) match {
case (p1, ..., pN) if G => B
...
case _ if G => B
}
```

to

```
val t1 = e1
... val tN = eN

t1 | ... | TN match {
case p1 | ... | pN if G => B
...
case _ if G => B
}
```

Where `|` is like magic: its constructor is a no-op, and its accessors are
local variable accesses (the `ti`).

The optimization supports scrutinees of the shape:
- `new TupleN(e1,...,eN)`
- `TupleN.apply(e1,...,eN)`

This is not allowed (can't omit the effect of the `???`)
- `{???; TupleN}.apply(e1,...,eN)` // !isQualifierSafeToElide(fun)

Built on top of a bunch of refactorings:
- SubstOnlyTreeMaker can subst >1 binders
- peel off a few Layers of Abstraction
- simplify `BoundTree`
- move `makers` to ExtractorCall
- pull out scrutineeFor
- Scrutinee encapsulates its sym & tree
- Simplify generation of default case

Already incorporates over-the-shoulder review by @retronym. Review by @retronym.

adriaanm committed:
[refactor] `Scrutinee` encapsulates its symbol & tree

Most of the changes can be summarized as:
  - `s/scrut: Tree, scrutSym: Symbol/scrutinee: Scrutinee/`
  - `s/prevBinder/scrutinee/`
  - `s/scrutSym/scrutinee.sym/`
  - `s/binder/scrutinee.sym/`

A later commit will introduce a scrutinee used for detuplification.

Variable numbering changed slightly, so of course `inline-ex-handlers`
needed to be updated.

adriaanm committed:
[refactor] Simplify generation of default case

Both `scrutSym` and `matchFailGen` are known when calling `codegen.matcher`,
and `codegen.matcher` simply applies `matchFailGen` to `REF(scrutSym)`.
So, just pass the result of the application to `codegen.matcher`,
rather than the function and the argument.

Also, renamed `matchFailGen` to `defaultCase`.

adriaanm committed:
[refactor] Lift `scrutineeFor`; `makers` to `ExtractorCall`

A later commit will generalize `scrutineeFor`
to detect a tupled scrutinee that we can optimize away.

Regarding pulling out `makers` from `extractorStep`:
the `import extractor.treeMaker` was a nice hint that, maybe,
the whole method would make more sense on the class of `extractor`.

Inlined some single-use private methods, reduced scope of others.

Also, `s/subBoundTrees/subPatterns/`.

adriaanm committed:
[refactor] Simplify `BoundTree`

Now called `PatternTranslation`, some more gratuitous renaming.
Inline `translatePattern` and `newBoundTree`.

adriaanm committed:
[refactor] Peel off a few Layers of Abstraction

The essence of the translation can now actually be understood
by looking at `def translate` in `class PatternTranslation`.

adriaanm committed:
SI-8790 Detuplify match on tuple under -Xexperimental

Optimize

```
(e1,..., eN) match {
  case (p1, ..., pN) if G => B
  ...
  case _ if G => B
}
```

to

```
val t1 = e1
...
val tN = eN

t1 | ... | TN match {
  case p1 | ... | pN if G => B
  ...
  case _ if G => B
}
```

Here, `|` is truly a magic class: its constructor is a no-op,
and its accessors are local variable accesses (the `ti`).

The optimization supports scrutinees of the shape:
  - `(e1,...,eN)` (by the time we're looking at it, desugared to the next one)
  - `TupleN.apply(e1,...,eN)`
  - `new TupleN(e1,...,eN)`

Stuff like this is not allowed (can't omit the effect of the `???`)
  - `{???; TupleN}.apply(e1,...,eN)` // !isQualifierSafeToElide(fun)

TODO: this needs to be tested more thoroughly before it can
be un-Xexperimentalized (although I think I fixed the breakage in CSE
that blocked the PR last time -- see next commit).

retronym commented:
I'd like to see some tests for:
- [x] `case t @ (a, b)` => t.isInstanceOf[Tuple2] // checking that no optimization was attempted
- [x] `{ ??? ; Tuple2}(a, b) match { case _ => }` // expect throw

I'd also like to see at least one or two tests to show that the optimization kicks in. Perhaps the tightest approach would be to borrow @lrytz's infrastructure from optimizer tests: e.g `test/files/junit/scala/tools/nsc/backend/jvm/opt/EmptyExceptionHandlersTest.scala`

adriaanm commented:
I have partial coverage for these in 
https://github.com/scala/scala/pull/4376/files#diff-0f83302065973206a21855b2ce946679. I'll add one for the instanceOf example.

adriaanm commented:
The earlier test failure was spurious. I hit rebuild on it.
Community build: https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/147/console

adriaanm commented:
For shits & giggles, with the optimization always-on: https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/148/console

adriaanm commented:
/nothingtoseehere -- only aggregator was failing, all actual tasks green

adriaanm commented:
always-on community build: https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/152/console

adriaanm committed:
Avoid a source of NPE in erasure due to tree reuse

The test actually doesn't do anything, because CSE
is disabled for `TupleMatchScrutinee`s,
which break CSE beyond the fix in this commit.

adriaanm commented:
@retronym I figured why erasure was NPEing. Queued another community build to see how we're doing now. I hope this will be ready to go now regardless, though.

adriaanm commented:
Yeah, I had to disable CSE when we detuplify because they interact badly.

adriaanm commented:
Community build is blue! I'll clarify test

soc commented:
Great change! I'm not really qualified to say anything, but I wonder whether `TupledScrutinee` works correctly for subclasses of Tuple? Extending Tuple is deprecated, but bad things could still happen until it is removed.

adriaanm commented:
Yep, sneaky subclasses would break under this optimization, which is why the scrutinee must syntactically be a tuple. I think that should keep us safe, right?

soc commented:
I agree, yes!

adriaanm committed:
[refactor] simplify scrutinee-based suppression of patmat analysis

adriaanm commented:
rebased.

adriaanm commented:
Don't know what to make of https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/568/ -- retrying at https://scala-ci.typesafe.com/job/scala-2.11.x-validate-test/571

retronym commented:
@adriaanm That assertion error is the same one that @gkossakowski's encountered while working on the SBT build. When we also saw it during PR validation, I added (#4397) an extended diagnostic error to the assertion in the hope we would see it again.

Now, to make sense of this:

```
[quick.compiler] error: symbol value x2#803320 does not exist in scala.tools.nsc.typechecker.ContextErrors$TyperContextErrors$TyperErrorGen.AdaptTypeError, which contains locals value x1#803317,value foundType#508052,value req#508050,value found#508049,value tree#508048
[quick.compiler] error: scala.reflect.internal.FatalError: symbol value x2#803320 does not exist in scala.tools.nsc.typechecker.ContextErrors$TyperContextErrors$TyperErrorGen.AdaptTypeError, which contains locals value x1#803317,value foundType#508052,value req#508050,value found#508049,value tree#508048
```

I've raised a critical ticket for this: https://issues.scala-lang.org/browse/SI-9264

adriaanm commented:
/synch

adriaanm commented:
rebased on 2.11.x with better diagnostic

sjrd commented:
How does this behave on a `match` without default case? For example, removing the `case _ =>` from the test case:

``` scala
  def a(x1: Any, x2: Any): Int = (x1, x2) match {
    case (s1: String, s2: String) => return 1
    case (i1: Int, s2)            => return 2
    case (s1, s2: String)         => return 3
  }
```

Currently, this automatically adds the following default case:

``` scala
case tup => throw new MatchError(tup)
```

which, AFAIU, would prevent this optimization from applying because it captures `tup`?

Instead, the above could generate a different synthetic default case:

``` scala
case _ => throw new MatchError((x1, x2))
```

Since the pair was just created to do the `match`, this is semantically equivalent, and would allow the optimization to apply.

Instead, this can always be done on a case-by-case basis: every case that captures the just-constructed tuple, of the form:

``` scala
case tup if G => body
```

would be rewritten as:

``` scala
case _ if G => val tup = (x1, x2); body
```

thereby avoiding the capture of the tuple instance created for the match, without changing the semantics. And then the optimization would apply. For those branches that captured `tup`, of course they will still do the allocation. But the branches that _don't_ won't need the allocation, and these are supposed to be the good paths.
(This is complicated if `G` refers to `tup`, of course, in which case that might not be possible in the general case. But usually it doesn't.)

This is precisely the only problem we have with tuple-matches in the Scala.js optimizer. If there is a default case, our optimizer is already able to do all that this PR brings, simply as a consequence of stack allocation. But the synthetic default case that puts `tup` in a `MatchError` kills us so far.

adriaanm commented:
Thanks for the suggestions! I added an empty commit with your note for when I pick this up again -- hopefully in a week or two.
